providers/amazon/tests/unit/amazon/aws/bundles/test_s3.py

```diff
@@ -0,0 +1,305 @@
+            prefix=S3_BUCKET_PREFIX,
+            bucket_name=S3_BUCKET_NAME,
+        )
+        # initialize succeeds, with correct prefix and bucket
+        bundle.initialize()
+        assert bundle.s3_hook.region_name == AWS_CONN_ID_REGION
+
+        bundle = S3DagBundle(
+            name="test",
+            aws_conn_id=AWS_CONN_ID_WITH_REGION,
+            prefix="",
+            bucket_name=S3_BUCKET_NAME,
+        )
+        # initialize succeeds, with empty prefix
+        bundle.initialize()
+        assert bundle.s3_hook.region_name == AWS_CONN_ID_REGION
+
+    def _upload_fixtures(self, bucket: str, fixtures_dir: str) -> None:
+        client = boto3.client("s3")
+        fixtures_paths = [
+            os.path.join(path, filename) for path, _, files in os.walk(fixtures_dir) for filename in files
+        ]
+        for path in fixtures_paths:
+            key = os.path.relpath(path, fixtures_dir)
+            client.upload_file(Filename=path, Bucket=bucket, Key=key)
+
+    @pytest.mark.db_test
+    def test_refresh(self, s3_bucket, s3_client, caplog, cap_structlog):
+        caplog.set_level(logging.ERROR)
+        caplog.set_level(logging.DEBUG, logger="airflow.providers.amazon.aws.bundles.s3.S3DagBundle")
+        caplog.set_level(
+            logging.DEBUG, logger="airflow.task.hooks.airflow.providers.amazon.aws.hooks.s3.S3Hook"
+        )
+        bundle = S3DagBundle(
+            name="test",
+            aws_conn_id=AWS_CONN_ID_WITH_REGION,
+            prefix=S3_BUCKET_PREFIX,
+            bucket_name=S3_BUCKET_NAME,
+        )
+
+        bundle.initialize()
+        # dags are downloaded once by initialize and once with refresh called post initialize
+        assert cap_structlog.text.count("Downloading DAGs from s3") == 1
+        self.assert_log_matches_regex(
+            caplog=caplog,
+            level="DEBUG",
+            regex=rf"Downloaded.*{S3_BUCKET_PREFIX}.*subproject1/dag_a.py.*{bundle.s3_dags_dir.as_posix()}.*subproject1/dag_a.py.*",
+        )
+
+        s3_client.put_object(Bucket=s3_bucket.name, Key=S3_BUCKET_PREFIX + "/dag_03.py", Body=b"test data")
+        bundle.refresh()
+        assert cap_structlog.text.count("Downloading DAGs from s3") == 2
+        self.assert_log_matches_regex(
+            caplog=caplog,
+            level="DEBUG",
+            regex=rf"Local file.*/{bundle.name}/subproject1/dag_a.py.*is up-to-date with S3 object.*{S3_BUCKET_PREFIX}.*subproject1/dag_a.py.*",
+        )
+        self.assert_log_matches_regex(
+            caplog=caplog,
+            level="DEBUG",
+            regex=rf"Downloaded.*{S3_BUCKET_PREFIX}.*dag_03.py.*/{bundle.name}/dag_03.py",
+        )
+        assert bundle.s3_dags_dir.joinpath("dag_03.py").read_text() == "test data"
+        bundle.s3_dags_dir.joinpath("dag_should_be_deleted.py").write_text("test dag")
+        bundle.s3_dags_dir.joinpath("dag_should_be_deleted_folder").mkdir(exist_ok=True)
+        s3_client.put_object(
```

Comment:
We should probably refactor this so the "syncing" stuff moved to the hook is tested with the hook tests instead.

Reply:
added syncing test to hook